larq/optimizers_v2.py

```diff
@@ -58,13 +58,23 @@ def _create_slots(self, var_list):
                 self.add_slot(var, "m")
 
     def apply_gradients(self, grads_and_vars, name=None):
+        # convert from zip object so we can loop multiple times:
+        grads_and_vars = list(grads_and_vars)
+
         bin_grads_and_vars = [(g, v) for g, v in grads_and_vars if self.is_binary(v)]
         fp_grads_and_vars = [(g, v) for g, v in grads_and_vars if not self.is_binary(v)]
 
         bin_train_op = super().apply_gradients(bin_grads_and_vars, name=name)
-        fp_train_op = self.fp_optimizer.apply_gradients(fp_grads_and_vars, name=name)
 
-        return tf.group(bin_train_op, fp_train_op, name="train_with_bop")
+        # only include real-valued update op if necessary (to avoid errors
+        # in distributed setting).
+        if fp_grads_and_vars:
```

Comment:
This should be fixed in https://github.com/tensorflow/tensorflow/commit/6b346a7b3fdb4d7db504f46e7b045694edee68a8#diff-39c47d5e727df3ac37c8c60f072a1e68

Though I am not sure if the commit is already included in TF 2.0.0

Reply:
Actually I am not sure anymore if we should fix this bug here, since in practice the list will probably never be empty. But I am OK to include this fix until TF 2.0.x is released.

Reply:
You're right, this fixes the same issue. As this situation is exceedingly rare (pure binary model in distributed context), and generates an error instead of silent weird behaviour, I'll take our fix out.